File: docs/research/uniffi-did-spike/swift/Smoke.swift
import Foundation

@main
struct Smoke {
    static func main() throws {
        let did = try parseDid(value: "did:example:123")
        precondition(did.value == "did:example:123")
        precondition(did.method == "example")
        precondition(did.methodSpecificId == "123")

        let didUrl = try parseDidUrl(value: "did:example:123/path?service=agent#key-1")
        precondition(didUrl.value == "did:example:123/path?service=agent#key-1")
        precondition(didUrl.did == "did:example:123")
        precondition(didUrl.path == "/path")
        precondition(didUrl.query == "service=agent")
        precondition(didUrl.fragment == "key-1")

        let callerText = "did:EXAMPLE:do-not-reflect"
        do {
            _ = try parseDid(value: callerText)
            preconditionFailure("invalid DID was accepted")
        } catch let error as DidBindingError {
            precondition(error == .InvalidDid)
            precondition(!error.localizedDescription.contains(callerText))
        }

        let oversized = "did:example:" + String(repeating: "a", count: 4_096)
        do {
            _ = try parseDidUrl(value: oversized)
            preconditionFailure("oversized DID URL was accepted")
        } catch let error as DidBindingError {
            precondition(error == .InvalidDidUrl)
        }

        print("swift-smoke: passed")
    }
}

File: docs/research/uniffi-did-spike/kotlin/src/main/kotlin/org/hyperledger/identus/spike/Smoke.kt
package org.hyperledger.identus.spike

fun main() {
    val did = parseDid("did:example:123")
    check(did.value == "did:example:123")
    check(did.method == "example")
    check(did.methodSpecificId == "123")

    val didUrl = parseDidUrl("did:example:123/path?service=agent#key-1")
    check(didUrl.value == "did:example:123/path?service=agent#key-1")
    check(didUrl.did == "did:example:123")
    check(didUrl.path == "/path")
    check(didUrl.query == "service=agent")
    check(didUrl.fragment == "key-1")

    val callerText = "did:EXAMPLE:do-not-reflect"
    try {
        parseDid(callerText)
        error("invalid DID was accepted")
    } catch (error: DidBindingException.InvalidDid) {
        check(!error.message.orEmpty().contains(callerText))
    }

    val oversized = "did:example:" + "a".repeat(4_096)
    try {
        parseDidUrl(oversized)
        error("oversized DID URL was accepted")
    } catch (_: DidBindingException.InvalidDidUrl) {
    }

    println("kotlin-smoke: passed")
}
